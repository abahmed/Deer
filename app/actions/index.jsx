// Local libraries
import { ACTION_TYPES } from './types';

/*
*  All notes related actions including payloads, if any
*/

// fetchAllNotes - Retrieve notes from the store

export function fetchAllNotes() {
  return {
    type: ACTION_TYPES.FETCH_ALL_NOTES,
  };
}

// addNote - Add/Save a new note to the store
export function addNote(payload) {
  return {
    type: ACTION_TYPES.ADD_NOTE,
    payload: payload,
  };
}

// const ACTIONS = {
//   addNote: addNote,
//   fetchAllNotes: fetchAllNotes
// }

// export default ACTIONS;

