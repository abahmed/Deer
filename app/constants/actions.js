export const ACTIONS = {
  GET_NEXT_LANG: 'SET_NEXT_LANG',
  TOGGLE_FADE: 'TOGGLE_FADE',
  UPDATE_EDITOR_STATE: 'UPDATE_EDITOR_STATE',
  FETCH_ALL_NOTES: 'FETCH_ALL_NOTES',
  UPDATE_NOTE_LIST: 'UPDATE_NOTE_LIST',
  SET_ACTIVE_NOTE: 'SET_ACTIVE_NOTE',
  UPDATE_NEW_NOTE_STATE: 'UPDATE_NEW_NOTE_STATE',
  UPDATE_SAVE_NOTE_STATE: 'UPDATE_SAVE_NOTE_STATE',
  ADD_NOTE: 'ADD_NOTE'
}
