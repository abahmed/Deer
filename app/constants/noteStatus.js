export const NOTE_STATUS = {
  NO_OPERATION: 'NO_OPERATION',
  SAVING_NOTE: 'SAVING_NOTE',
  NOTE_SAVE_SUCCESS: 'NOTE_SAVE_SUCCESS',
  NOTE_SAVE_FAIL: 'NOTE_SAVE_FAIL',
  LOADING_NOTE: 'LOADING_NOTE',
  NOTE_LOAD_SUCCESS: 'NOTE_LOAD_SUCCESS',
  NOTE_LOAD_FAIL: 'NOTE_LOAD_FAIL'
}
