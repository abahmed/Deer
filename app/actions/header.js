import { createAction } from 'redux-actions'
import { ACTIONS } from '../constants/actions'

// Used for updating state of adding new note button (e.g. enabled or disabled).
export const setNewNoteDisabled = createAction(ACTIONS.UPDATE_NEW_NOTE_STATE)

// Used for updating state of save button (e.g. enabled or disabled).
export const setSaveDisabled = createAction(ACTIONS.UPDATE_SAVE_NOTE_STATE)

// Used for updating state of deleteing note button (e.g. enabled or disabled).
export const setDeleteDisabled = createAction(ACTIONS.UPDATE_DELETE_NOTE_STATE)
