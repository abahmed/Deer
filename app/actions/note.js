import { createAction } from 'redux-actions'
import { ACTIONS } from '../constants/actions'
import { fetchNotes } from './../db'

// Used for adding a new note.
export const addNewNote = createAction(ACTIONS.ADD_NOTE)

// Used for updating noteList with fetched notes from database.
export const updateNoteList = createAction(ACTIONS.UPDATE_NOTE_LIST)

// Used for setting the id of selected note.
export const setActiveNote = createAction(ACTIONS.SET_ACTIVE_NOTE)

// Used for updating title of active note.
export const updateNoteTitle = createAction(ACTIONS.UPDATE_NOTE_TITLE)

// Used for fetching all notes from database.
export const fetchAllNotes = () => (dispatch, getState) => {
  fetchNotes().then((result) => {
    // Update only if there are notes.
    if (result.total_rows > 0) {
      dispatch(updateNoteList(result.rows))
    }
  }).catch((err) => {
    console.error(err)
  })
}
