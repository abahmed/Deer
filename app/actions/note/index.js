import { createAction } from 'redux-actions'
import { ACTIONS } from '../../constants/actions'
import { addNote, fetchNotes, removeNote } from '../../utils/db'
import logger from 'electron-log'

/** Used for adding a new note. */
const addNewNote = createAction(ACTIONS.ADD_NOTE)

/** Used for updating notes with fetched notes from database. */
const updateNotesList = createAction(ACTIONS.NOTES_FETCH_SUCCESS)

/** Used for setting the index of selected note. */
const setSelectedNoteID = createAction(ACTIONS.SET_SELECTED_NOTE_ID)

/** Used for updating title, content and modified values for the selected note. */
const editSelectedNote = createAction(ACTIONS.EDIT_SELECTED_NOTE)

/** Used for updating rev of the selected note. */
const updateSelectedNoteRev = createAction(ACTIONS.UPDATE_SELECTED_NOTE_REV)

/** Used for deleting a note from noteList. */
const deleteSelectedNote = createAction(ACTIONS.DELETE_SELECTED_NOTE)

/** Async method, Used for fetching all notes from database. */
const fetchAllNotes = () => (dispatch) => {
  fetchNotes().then((result) => {
    // Update only if there are notes.
    if (result.total_rows > 0) {
      dispatch(updateNotesList(result.rows))
    }
  }).catch((err) => {
    logger.error('Unable to fetch notes: ' + JSON.stringify(err))
  })
}

/**
 * Async method, used for saving note to database.
 */
const createNote = () => (dispatch) => {
  const modified = Date.now()
  addNote(undefined, undefined, undefined, modified).then((result) => {
    if (result.ok) {
      dispatch(addNewNote({
        id: result.id,
        rev: result.rev,
        title: '',
        content: '',
        modified: modified
      }))
      dispatch(setSelectedNoteID(result.id))
    }
  }).catch((err) => {
    logger.error('Unable to add new note: ' + JSON.stringify(err))
  })
}

/**
 * Async method, used for updating note in database.
 */
const saveSelectedNote = () => (dispatch, getState) => {
  const state = getState().noteReducer
  const selectedNoteID = state.get('selectedNoteID')
  const selectedNote = state.get('notes')[selectedNoteID]
  addNote(
    selectedNoteID,
    selectedNote.title,
    selectedNote.content,
    selectedNote.modified,
    selectedNote.rev
  ).then((result) => {
    if (result.ok) {
      // Update note rev to avoid conflicts while the next time for saving
      // this note.
      dispatch(updateSelectedNoteRev(result.rev))
    }
  }).catch((err) => {
    logger.error('Unable to update note: ' + JSON.stringify(err))
  })
}


/** Async method, used for removing active note from database. */
const removeSelectedNote = () => (dispatch, getState) => {
  const state = getState().noteReducer
  const selectedNoteID = state.get('selectedNoteID')
  const selectedNote = state.get('notes')[selectedNoteID]
  removeNote(selectedNoteID, selectedNote.rev).then((result) => {
    dispatch(deleteSelectedNote())
  }).catch((err) => {
    logger.error('Unable to remove note ' + JSON.stringify(err))
  })
}

export {
  updateNotesList,
  addNewNote,
  setSelectedNoteID,
  editSelectedNote,
  updateSelectedNoteRev,
  fetchAllNotes,
  createNote,
  saveSelectedNote,
  deleteSelectedNote,
  removeSelectedNote
}
