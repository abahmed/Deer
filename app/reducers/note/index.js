import { ACTIONS } from '../../constants/actions'
import logger from 'electron-log'
import { Map, merge } from 'immutable'

const INITIAL_STATE = Map({
  notes: {},
  selectedNoteID: ACTIONS.NOT_SELECTED_NOTE,
  searchNotes: []
})

const createNote = (title, content, modified, rev) => {
  return {
    title: title,
    content: content,
    modified: modified || '',
    rev: rev || ''
  }
}

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case ACTIONS.NOTES_FETCH_SUCCESS:
      let notes = {}
      action.payload.forEach(note => {
        notes[note.doc._id] = createNote(
          note.doc.title,
          note.doc.content,
          note.doc.modified,
          note.doc._rev
        )
      })
      return state.set('notes', merge(state.get('notes'), notes))
    case ACTIONS.SET_SELECTED_NOTE_ID:
      const noteID = action.payload
      if (noteID === state.get('selectedNoteID')) {
        return state
      }
      else if (noteID === ACTIONS.NOT_SELECTED_NOTE &&
          state.get('selectedNoteID') !== ACTIONS.NOT_SELECTED_NOTE) {
        return state.set('selectedNoteID', ACTIONS.NOT_SELECTED_NOTE)
      }
      else if (!state.get('notes').hasOwnProperty(noteID)) {
        logger.error('Selected note with invalid id: ' + noteID)
        return state
      }
      return state.set('selectedNoteID', noteID)
    case ACTIONS.ADD_NOTE:
      const newNote = action.payload
      if (state.get('notes').hasOwnProperty(newNote.id)) {
        logger.error('This note already exists: ' + newNote.id)
        return state
      }
      return state.setIn(['notes', newNote.id], createNote(
        newNote.title,
        newNote.content,
        newNote.modified,
        newNote.rev
      ))
    case ACTIONS.EDIT_SELECTED_NOTE:
      const selectedNoteID = state.get('selectedNoteID')
      if (selectedNoteID === ACTIONS.NOT_SELECTED_NOTE) {
        logger.error('There is no selected note')
        return state
      }
      const note = state.get('notes')[selectedNoteID]
      const newValues = action.payload
      if (newValues.content === note.content) {
        return state
      }
      return state.setIn(['notes', state.get('selectedNoteID')], createNote(
        newValues.title,
        newValues.content,
        newValues.modified,
        note.rev
      ))
    case ACTIONS.UPDATE_SELECTED_NOTE_REV:
      if (state.get('selectedNoteID') === ACTIONS.NOT_SELECTED_NOTE) {
        logger.error('There is no selected note')
        return state
      }
      return state.setIn(['notes', state.get('selectedNoteID'), 'rev'],
                         action.payload)
    case ACTIONS.DELETE_SELECTED_NOTE:
      if (state.get('selectedNoteID') === ACTIONS.NOT_SELECTED_NOTE) {
        logger.error('There is no selected note')
        return state
      }
      return state.removeIn(['notes', state.get('selectedNoteID')])
                  .set('selectedNoteID', ACTIONS.NOT_SELECTED_NOTE)
    case ACTIONS.NOTES_SEARCH_UPDATE:
      let searchNotes = []
      action.payload.forEach(note => {
        searchNotes.push(note.id)
      })
      if (state.get('searchNotes') === searchNotes)
        return state

      return state.set('searchNotes', searchNotes)
    default:
      return state
  }
}
