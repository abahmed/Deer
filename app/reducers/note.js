import uuidv4 from 'uuid/v4'
import { EditorState } from 'draft-js'
import { ACTIONS } from '../constants/actions'
import { NOTE_STATUS } from '../constants/noteStatus'
import logger from 'electron-log'

const INITIAL_STATE = {
  activeNoteIndex: -1,
  activeNoteState: EditorState.createEmpty(),
  noteStatus: NOTE_STATUS.NO_OPERATION,
  notes: []
}

// Helper method, Updates a field with a newValue of an element in notes array
// without altering provided state if there is a change in value, otherwise
// returns provided state.
const _updateNoteEntry = (state, field, newValue) => {
  // Just return state as there is no active note.
  if (state.activeNoteIndex < 0 ||
      state.activeNoteIndex >= state.notes.length) {
    logger.warn('UPDATE_NOTE_TITLE action is fired with out of range' +
                 ' activeNoteIndex')
    return state
  }

  const currentNote = Object.create(state.notes[state.activeNoteIndex])

  // Just return the state if there is not a change.
  if (currentNote[field] === newValue) { return state }

  currentNote[field] = newValue
  return {
    ...state,
    notes: [
      ...state.notes.slice(0, state.activeNoteIndex),
      currentNote,
      ...state.notes.slice(state.activeNoteIndex + 1)
    ]
  }
}

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case ACTIONS.UPDATE_NOTE_LIST:
      return {
        ...state,
        notes: action.payload.map(note => {
          return {
            id: note.doc._id,
            rev: note.doc._rev,
            title: note.doc.title
          }
        })
      }
    case ACTIONS.SET_ACTIVE_NOTE_INDEX:
      return {
        ...state,
        activeNoteIndex: action.payload
      }
    case ACTIONS.ADD_NOTE:
      return {
        ...state,
        notes: [...state.notes, {
          id: uuidv4(),
          rev: '',
          title: ''
        }],
        activeNoteIndex: state.notes.length,
        activeNoteState: EditorState.createEmpty()
      }
    case ACTIONS.UPDATE_NOTE_TITLE:
      return _updateNoteEntry(state, 'title', action.payload)
    case ACTIONS.UPDATE_NOTE_REV:
      return _updateNoteEntry(state, 'rev', action.payload)
    case ACTIONS.UPDATE_ACTIVE_NOTE_STATE:
      return {
        ...state,
        activeNoteState: action.payload
      }
    case ACTIONS.SET_NOTE_STATUS:
      if (!NOTE_STATUS.hasOwnProperty(action.payload)) {
        logger.warn('Trying to set unsupported noteStatus')
        return state
      }

      return {
        ...state,
        noteStatus: action.payload
      }
    case ACTIONS.LOAD_NOTE_CONTENT:
      return {
        ...state,
        activeNoteState: EditorState.createWithContent(action.payload)
      }
    default:
      return state
  }
}
