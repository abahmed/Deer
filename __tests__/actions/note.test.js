import configureStore from 'redux-mock-store'

// Actions to be tested
import * as noteActions from '../../app/actions/note'
import { NOTE_STATUS } from '../../app/constants/noteStatus'

const mockStore = configureStore()
const store = mockStore()

describe('note_actions', () => {
  beforeEach(() => {
    store.clearActions()
  })

  describe('addNewNote', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(noteActions.addNewNote())
      expect(store.getActions()).toMatchSnapshot()
    })
  })

  describe('updateNoteList', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(noteActions.updateNoteList())
      expect(store.getActions()).toMatchSnapshot()
    })
  })

  describe('setActiveNoteIndex', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(noteActions.setActiveNoteIndex(3))
      expect(store.getActions()).toMatchSnapshot()
    })
  })

  describe('updateNoteTitle', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(noteActions.updateNoteTitle('Hello world'))
      expect(store.getActions()).toMatchSnapshot()
    })
  })

  describe('updateActiveNoteState', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(noteActions.updateActiveNoteState({}))
      expect(store.getActions()).toMatchSnapshot()
    })
  })

  describe('updateNoteRev', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(noteActions.updateNoteRev('a123-a123'))
      expect(store.getActions()).toMatchSnapshot()
    })
  })

  describe('setNoteStatus', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(noteActions.setNoteStatus(NOTE_STATUS.SAVING_NOTE))
      expect(store.getActions()).toMatchSnapshot()
    })
  })

  describe('loadNoteContent', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(noteActions.loadNoteContent({}))
      expect(store.getActions()).toMatchSnapshot()
    })
  })

  describe('deleteNoteFromList', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(noteActions.deleteNoteFromList(3))
      expect(store.getActions()).toMatchSnapshot()
    })
  })
})
