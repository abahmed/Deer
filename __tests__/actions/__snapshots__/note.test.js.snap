// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`note_actions addNewNote Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "type": "ADD_NOTE",
  },
]
`;

exports[`note_actions deleteNoteFromList Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": 3,
    "type": "DELETE_NOTE_FROM_LIST",
  },
]
`;

exports[`note_actions loadNoteContent Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "LOAD_NOTE_CONTENT",
  },
]
`;

exports[`note_actions setActiveNoteIndex Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": 3,
    "type": "SET_ACTIVE_NOTE",
  },
]
`;

exports[`note_actions setNoteStatus Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": "SAVING_NOTE",
    "type": "SET_NOTE_STATUS",
  },
]
`;

exports[`note_actions updateActiveNoteState Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "UPDATE_ACTIVE_NOTE_STATE",
  },
]
`;

exports[`note_actions updateNoteList Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "type": "UPDATE_NOTE_LIST",
  },
]
`;

exports[`note_actions updateNoteRev Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": "a123-a123",
    "type": "UPDATE_NOTE_REV",
  },
]
`;

exports[`note_actions updateNoteTitle Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": "Hello world",
    "type": "UPDATE_NOTE_TITLE",
  },
]
`;
